### ISSUE TITLE
Handle assertions during JVM startup

### ISSUE THREAD

**Issue by FabienSe** [jpype-project/jpype#1170]
Hello all.

I have a python application which was working great but today the startJVM is crashing :/
I did not change anything in my application. I do not see any logs.

By looking at the windows logs I can see the python.exe processus is crashing because of ntdll.dll, version : 10.0.19041.3636.
located at C:\WINDOWS\SYSTEM32\ntdll.dll.
After debug, I can see it is crashing in the startJVM call I perform.

Here is the call I perform

`startJVM(str(jvm_path), "-ea", f"-Djava.library.path={resources_path}", classpath=classpaths, interrupt=False)`
I need to start the JVM with custom jar loaded with it.

I tried to start it with `startJVM` with no parameters and it seems it is working.

I do not have any idea on how to fix this problem and it is crazy that it is crashing now without any changes to my code.

Hopefully you can help me with my problem.

Thanks a lot for your help.

**Comment by Thrameos** [comment]
At best I can make some guess.   It must be some change in the jvm_path or the dll that is located at the point.   Perhaps it something in the script is now doing mixed architectures (32 vs 64).   I would try to break down each step from the one which works to the one that doesn’t.

Alternatively it may be something that the JVM is loading indirectly like something in the resources path.  Look for anything in those paths that changed in the time.    Were it my site, where the system admin runs nightly “security patches”, I would assume that the admin pulled the rug out from me by changing the jdk or libraries.

Beyond these general tips I don’t think JPype devels can be much help as it is clearly a site install issue.

**Comment by Thrameos** [comment]
No.  Though it is not very surprising that “-ea” would cause a problem.   That enables assertions so that means something either in JPype, Java, or the jars loaded has an assert statement which is failing.   Not sure how to find the failed assertion unless we can get some logs.

Can you try using the gdb instructions to capture a log of the failure?   Either the traceback log or the crash dump from the JVM should show you the code that made the assert.

**Comment by Thrameos** [comment]
Looking at the documentation it appears that the assertion is sending up an exception.   If it happens while the JVM is still getting going, then most likely sending us the flaming death.   I think we should just put an issue to handle “java.lang.Error” during startup.

**Comment by Thrameos** [comment]
We can use this one for now.   Just need to rename it to the required task for now.   Something like “Handle assertions during JVM startup”

The issue is that the assertion is failing some place, and it is likely bubbling up at a point in time when the system is vulnerable because we don’t yet have all the resources in place the handle it.

The solution will be to deliberately add a bad assertion at different points in the JPype support jar file.   We then force the assertion to fail which will reproduce the crash.   We then redirect the crash to a normal Python exception so the user can see it.   It doesn’t mean that JPype will run as the best we can do is provide diagnostics.   This will need to be repeated at different places in the start up code as we don’t’ know where it is currently happening so we would have to guard all points.

**Comment by Thrameos** [comment]
Will revisit at some point.

**Comment by Thrameos** [comment]
Did a focused deep dive on this today. Summary of findings:

- There are no `assert` statements anywhere in JPype's own Java support code (`org.jpype.jar`), so nothing in JPype itself trips on `-ea` today.
- To test the "assertion firing during startup crashes/isn't handled" hypothesis directly, I injected a deliberate `assert false` at the very first line of `JPypeContext.createContext` — the earliest possible point in JPype's own Java bootstrap, before any of its own `try`/`catch` even starts — and rebuilt from a clean checkout.
- Running with `-ea` against that build, the failure surfaces as a clean Python exception (`SystemError: java.lang.AssertionError: ...`) on Linux. No crash, no hang. JPype's JNI exception-checking (`check()`, invoked after essentially every native call in `jp_javaframe.cpp`) already correctly catches and converts a `Throwable` raised at the earliest moment of startup.

So the original ask here ("handle java.lang.Error during startup") already appears to work, at least on Linux, with the exception surfacing as a Throwable at the very first opportunity. That points away from "JPype has an unguarded JNI call during startup" as the explanation for the original crash.

The reporter's actual crash signature (WinDbg: `ntdll!RtlpFreeHeapInternal`, `HEAP_CORRUPTION_ACTIONABLE_BlockNotBusy_DOUBLE_FREE`) is a distinct native heap-corruption failure mode, not a Java exception propagation issue - and it reproduced only on 6 specific "corporate managed" Windows machines and nowhere else, which points toward something in that environment (AV/EDR hooking `ntdll` heap functions, a JVM/CRT mismatch on those particular machines, etc.) rather than something fixable in JPype's startup code, matching the original "site install issue" guess.

Given this is a 2+ year old report tied to a single, no-longer-available machine fleet, with no further diagnostic info forthcoming, I'm closing this as best-effort / cannot-reproduce. Happy to reopen if anyone can attach a reproducible case (ideally on a machine we can get JVM crash logs / gdb output from) - the `-ea` + custom classpath combination is now easy to test if that happens.
